Lib/test/test_locale_coercion.py

```diff
@@ -0,0 +1,199 @@
+                     "C locale coercion disabled at build time")
+class LocaleOverrideTests(unittest.TestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        for target_locale, target_category, env_updates in _C_UTF8_LOCALES:
+            if _set_locale_in_subprocess(target_locale, target_category):
+                break
+        else:
+            raise unittest.SkipTest("No C-with-UTF-8 locale available")
+        cls.EXPECTED_COERCION_WARNING = CLI_COERCION_WARNING_FMT.format(
+            env_updates, target_locale
+        )
+
+    def _get_child_fsencoding(self, env_vars):
+        """Retrieves sys.getfilesystemencoding() from a child process
+
+        Returns (fsencoding, stderr_lines):
+
+        - fsencoding: a lowercase str value with the child's fsencoding
+        - stderr_lines: result of calling splitlines() on the stderr output
+
+        The child is run in isolated mode if the current interpreter supports
+        that.
+        """
+        cmd = "import sys; print(sys.getfilesystemencoding().lower())"
+        result, py_cmd = run_python_until_end(
+            "-c", cmd,
+            __isolated=True,
+            **env_vars
+        )
+        if not result.rc == 0:
+            result.fail(py_cmd)
+        # All subprocess outputs in this test case should be pure ASCII
+        child_fsencoding = result.out.decode("ascii").rstrip()
+        child_stderr_lines = result.err.decode("ascii").rstrip().splitlines()
+        return child_fsencoding, child_stderr_lines
+
+
+    def test_C_utf8_locale(self):
+        # Ensure the C.UTF-8 locale is accepted entirely without complaint
+        base_var_dict = {
+            "LANG": "",
+            "LC_CTYPE": "",
+            "LC_ALL": "",
+        }
+        for env_var in ("LC_ALL", "LC_CTYPE", "LANG"):
+            with self.subTest(env_var=env_var):
+                var_dict = base_var_dict.copy()
+                var_dict[env_var] = "C.UTF-8"
+                fsencoding, stderr_lines = self._get_child_fsencoding(var_dict)
+                self.assertEqual(fsencoding, "utf-8")
+                self.assertFalse(stderr_lines)
+
+
+    def _check_c_locale_coercion(self, expected_fsencoding, coerce_c_locale):
+        """Check the handling of the C locale for various configurations
+
+        Parameters:
+            expected_fsencoding: the encoding the child is expected to report
+            allow_c_locale: setting to use for PYTHONALLOWCLOCALE
+              None: don't set the variable at all
+              str: the value set in the child's environment
+        """
+        expected_warning = []
+        if coerce_c_locale != "0":
+            # Check C locale is coerced with a warning on stderr
+            expected_warning.append(self.EXPECTED_COERCION_WARNING)
+        base_var_dict = {
+            "LANG": "",
+            "LC_CTYPE": "",
+            "LC_ALL": "",
+        }
+        for env_var in ("LC_ALL", "LC_CTYPE", "LANG"):
```

Comment:
Done in https://github.com/python/cpython/pull/659/commits/d12b4126779b41c945c210124ba23b1c25ab62f5